# EDA(Exploratory Data Analysis) - group 11

## load packages

In [1]:
import numpy as np
import pandas as pd
import altair as alt

## load data

In [2]:
data = pd.read_csv("data/cricket_test.csv")
data = data.drop(columns = ["Unnamed: 0"])
data

game_id   season     team  over          batter batter_id       bowler  \
0      211028     2005  England     0  ME Trescothick  ea42ddb9        B Lee   
1      211028     2005  England     0  ME Trescothick  ea42ddb9        B Lee   
2      211028     2005  England     0        GO Jones  2e929b99        B Lee   
3      211028     2005  England     0        GO Jones  2e929b99        B Lee   
4      211028     2005  England     0        GO Jones  2e929b99        B Lee   
...       ...      ...      ...   ...             ...       ...          ...   
2323   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   
2324   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   
2325   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   
2326   255954  2006/07    India    19        SK Raina  1dc12ab9  RJ Peterson   
2327   255954  2006/07    India    19      KD Karthik  c03f1114  RJ Peterson   

     bowler_id     non_striker non_striker_id  ...  fielders_id  wicket_type  \
0     dd09ff8e        GO Jones       2e929b99  ...          NaN          NaN   
1     dd09ff8e        GO Jones       2e929b99  ...          NaN          NaN   
2     dd09ff8e  ME Trescothick       ea42ddb9  ...          NaN          NaN   
3     dd09ff8e  ME Trescothick       ea42ddb9  ...          NaN          NaN   
4     dd09ff8e  ME Trescothick       ea42ddb9  ...          NaN          NaN   
...        ...             ...            ...  ...          ...          ...   
2323  26ff4c29        SK Raina       1dc12ab9  ...          NaN          NaN   
2324  26ff4c29        SK Raina       1dc12ab9  ...          NaN          NaN   
2325  26ff4c29        SK Raina       1dc12ab9  ...          NaN          NaN   
2326  26ff4c29      KD Karthik       c03f1114  ...          NaN          NaN   
2327  26ff4c29        SK Raina       1dc12ab9  ...          NaN          NaN   

      runs_batter  runs_extras  runs_total  team_over over_ball inning  \
0               0            0           0  England_0         1      1   
1               1            0           1  England_0         2      1   
2               0            0           0  England_0         3      1   
3               0            0           0  England_0         4      1   
4               0            0           0  England_0         5      1   
...           ...          ...         ...        ...       ...    ...   
2323            6            0           6   India_19         1      2   
2324            0            0           0   India_19         2      2   
2325            1            0           1   India_19         3      2   
2326            1            0           1   India_19         4      2   
2327            1            0           1   India_19         5      2   

     runs_cumulative powerplay  
0                  0         1  
1                  1         1  
2                  1         1  
3                  1         1  
4                  1         1  
...              ...       ...  
2323            2490         0  
2324            2490         0  
2325            2491         0  
2326            2492         0  
2327            2493         0  

[2328 rows x 28 columns]

## Univariate Analysis

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2328 entries, 0 to 2327
Data columns (total 28 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   game_id          2328 non-null   int64 
 1   season           2328 non-null   object
 2   team             2328 non-null   object
 3   over             2328 non-null   int64 
 4   batter           2328 non-null   object
 5   batter_id        2328 non-null   object
 6   bowler           2328 non-null   object
 7   bowler_id        2328 non-null   object
 8   non_striker      2328 non-null   object
 9   non_striker_id   2328 non-null   object
 10  wides            2328 non-null   int64 
 11  noballs          2328 non-null   int64 
 12  legbyes          2328 non-null   int64 
 13  byes             2328 non-null   int64 
 14  wicket           2328 non-null   int64 
 15  player_out       136 non-null    object
 16  player_out_id    136 non-null    object
 17  fielders_name    91 non-null     

In [4]:
data.describe()

game_id         over        wides      noballs      legbyes  \
count    2328.000000  2328.000000  2328.000000  2328.000000  2328.000000   
mean   230039.682131     9.044674     0.032646     0.009021     0.033076   
std     14100.577340     5.697328     0.210917     0.115067     0.234949   
min    211028.000000     0.000000     0.000000     0.000000     0.000000   
25%    222678.000000     4.000000     0.000000     0.000000     0.000000   
50%    225271.000000     9.000000     0.000000     0.000000     0.000000   
75%    238195.000000    14.000000     0.000000     0.000000     0.000000   
max    255954.000000    19.000000     5.000000     2.000000     4.000000   

              byes       wicket  runs_batter  runs_extras   runs_total  \
count  2328.000000  2328.000000  2328.000000  2328.000000  2328.000000   
mean      0.007302     0.058419     1.196735     0.082045     1.278780   
std       0.141931     0.234585     1.565350     0.358655     1.553248   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     1.000000     0.000000     1.000000   
75%       0.000000     0.000000     1.000000     0.000000     1.000000   
max       4.000000     1.000000     6.000000     5.000000     8.000000   

         over_ball       inning  runs_cumulative    powerplay  
count  2328.000000  2328.000000      2328.000000  2328.000000  
mean     10.213058     1.839777      1096.707474     0.326031  
std       6.973901     0.366892       765.761312     0.468859  
min       1.000000     1.000000         0.000000     0.000000  
25%       4.000000     2.000000       387.000000     0.000000  
50%       9.000000     2.000000      1019.000000     0.000000  
75%      16.000000     2.000000      1721.250000     1.000000  
max      28.000000     2.000000      2493.000000     1.000000

The dataset contains 28 columns and 2328 observations related to cricket. The ratio of quantitative to categorical (ordinal and nominal) data is exactly 50:50. Note that the columns **player_out, player_out_id, fielder_name, fielders_id, and wicket_type** show a significant amount of missing values. Player_out and player_out_id may be missing due to inapplicability (e.g., no one is out), which results in fielder_name, fielder_id, and wicket_type being invalid in these observations. Thus, these missing data are likely classified as **missing at random(MAR)**. Given the complexity of imputing these values, they are unlikely to be included in our further analysis.

Note that, among the 28 columns, there were few quantitative attributes generated based on existing columns, and they are "over_ball", "inning", "runs_cumulative", and "powerplay". Detailed calculations of these attributes are in the cricsheet-json-parsing.ipynb file. 

In [5]:
def vis_bar(x_input):
    return alt.Chart(data).mark_bar().encode(
        x = x_input,
        y = "count()"
    ).properties(
        width = 150, 
        height = 150
    )

over = vis_bar("over")
wides = vis_bar("wides")
noballs = vis_bar("noballs")
legbyes = vis_bar("legbyes")
byes = vis_bar("byes")
wicket = vis_bar("wicket")
run_batter = vis_bar("runs_batter")
run_extras = vis_bar("runs_extras")
run_total = vis_bar("runs_total") 
over_ball = vis_bar("over_ball")
inning = vis_bar("inning")
runs_cumulative = vis_bar("runs_cumulative") 
powerplay = vis_bar("powerplay")

v1 = alt.hconcat(over, wides, noballs)
v2 = alt.hconcat(legbyes, byes, wicket)
v3 = alt.hconcat(run_batter, run_extras, run_total)
v4 = alt.hconcat(over_ball, inning, runs_cumulative, powerplay)
alt.vconcat(v1, v2, v3, v4)

alt.VConcatChart(...)

In [6]:
season = vis_bar("season")
team = vis_bar("team")

alt.hconcat(season, team)

alt.HConcatChart(...)

In [7]:
data["batter"].value_counts().reset_index()

batter  count
0          GC Smith    155
1    ME Trescothick    135
2         SB Styris     85
3        RT Ponting     85
4         DR Martyn     77
..              ...    ...
100        KD Mills      1
101       M Zondeki      1
102     Younis Khan      1
103      MP Vaughan      1
104      SI Mahmood      1

[105 rows x 2 columns]

In [8]:
data["bowler"].value_counts().reset_index()

bowler  count
0            SE Bond     66
1              B Lee     64
2          A Symonds     61
3           AR Adams     53
4           KD Mills     52
..               ...    ...
71           GB Hogg     13
72        TM Dilshan     12
73         LPC Silva      6
74         JH Kallis      6
75  DPMD Jayawardene      6

[76 rows x 2 columns]

In [9]:
data["non_striker"].value_counts().reset_index()

non_striker  count
0    ME Trescothick    154
1          GC Smith    122
2         SB Styris     82
3        RT Ponting     80
4         DR Martyn     71
..              ...    ...
99          J Lewis      2
100      MP Vaughan      2
101       DR Tuffey      1
102         J Botha      1
103    IDR Bradshaw      1

[104 rows x 2 columns]

Please be aware that all columns related to "ID" are not included in this analysis as they are primarily used for identification purposes and are unlikely to exhibit significant patterns. As observed, none of the quantitative attributes follow a normal distribution; most display right-skewed distributions(the only exception is "inning"). The "runs_cumulative" attribute seems to be evenly distributed. An interesting observation is found in the attribute "over" and "over_ball" where the frequency decreases with increasing values, suggesting further analysis. Regarding ordinal attributes, the dataset contains much more data from the 2005/06 season compared to other seasons, as well as more teams from "New Zealand," "South Africa," "Australia" and "England" than teams from other countries/regions. Lastly, concerning nominal attributes, GC Smith and ME Trescothick appear to have considerably more observations than other players in both the "non-striker" and "batter" roles.

In general, **none of the attributes in our dataset form a normal distribution**, with samples unevenly distributed across categorical attributes. It is important to remember this when training, testing, and interpreting the model, and to appropriately acknowledge this as a potential limitation.

## Multivariate Analysis - Correlation Analysis

In [10]:
corr_data = data[["over", "wides", "noballs", "legbyes", "byes", "wicket", "runs_batter", "runs_extras", "runs_total", "over_ball", "inning", "runs_cumulative", "powerplay"]]
corr_df = corr_data.corr()
corr_df

over     wides   noballs   legbyes      byes    wicket  \
over             1.000000 -0.079890  0.016428  0.003390 -0.009438  0.079074   
wides           -0.079890  1.000000 -0.012139 -0.021799 -0.007967 -0.038562   
noballs          0.016428 -0.012139  1.000000 -0.011041 -0.004035  0.012310   
legbyes          0.003390 -0.021799 -0.011041  1.000000 -0.007246 -0.035073   
byes            -0.009438 -0.007967 -0.004035 -0.007246  1.000000 -0.012818   
wicket           0.079074 -0.038562  0.012310 -0.035073 -0.012818  1.000000   
runs_batter      0.068739 -0.118384  0.023545 -0.107673 -0.039352 -0.190471   
runs_extras     -0.043225  0.566751  0.304862  0.635856  0.385004 -0.046777   
runs_total       0.059294  0.011560  0.094123  0.038311  0.049242 -0.202756   
over_ball       -0.070099  0.047858 -0.036669  0.026383  0.020570 -0.026262   
inning          -0.025973  0.006536 -0.006467 -0.008290  0.014226  0.008939   
runs_cumulative  0.059734  0.024780 -0.033427  0.020025  0.005587 -0.016577   
powerplay       -0.802432  0.066147  0.001222 -0.012110  0.015870 -0.036494   

                 runs_batter  runs_extras  runs_total  over_ball    inning  \
over                0.068739    -0.043225    0.059294  -0.070099 -0.025973   
wides              -0.118384     0.566751    0.011560   0.047858  0.006536   
noballs             0.023545     0.304862    0.094123  -0.036669 -0.006467   
legbyes            -0.107673     0.635856    0.038311   0.026383 -0.008290   
byes               -0.039352     0.385004    0.049242   0.020570  0.014226   
wicket             -0.190471    -0.046777   -0.202756  -0.026262  0.008939   
runs_batter         1.000000    -0.148173    0.973577   0.012810 -0.005700   
runs_extras        -0.148173     1.000000    0.081579   0.041803  0.001968   
runs_total          0.973577     0.081579    1.000000   0.022563 -0.005290   
over_ball           0.012810     0.041803    0.022563   1.000000  0.022921   
inning             -0.005700     0.001968   -0.005290   0.022921  1.000000   
runs_cumulative    -0.017133     0.019177   -0.012839   0.486317  0.487396   
powerplay          -0.028294     0.037638   -0.019824   0.035655  0.016512   

                 runs_cumulative  powerplay  
over                    0.059734  -0.802432  
wides                   0.024780   0.066147  
noballs                -0.033427   0.001222  
legbyes                 0.020025  -0.012110  
byes                    0.005587   0.015870  
wicket                 -0.016577  -0.036494  
runs_batter            -0.017133  -0.028294  
runs_extras             0.019177   0.037638  
runs_total             -0.012839  -0.019824  
over_ball               0.486317   0.035655  
inning                  0.487396   0.016512  
runs_cumulative         1.000000  -0.027310  
powerplay              -0.027310   1.000000

In [11]:
corr_ = corr_df.stack()
corr_ = corr_.reset_index()
corr_.columns = ['row', 'column', 'corr']
corr_

alt.Chart(corr_).mark_rect().encode(
    x = 'column',
    y = 'row',
    color = 'corr:Q',
    tooltip = 'corr:Q'
).properties(
    width = 400,
    height = 400
)

alt.Chart(...)

In [12]:
corr_table = corr_[corr_["row"] != corr_["column"]]
corr_table.sort_values("corr", ascending = False).head(15)

row           column      corr
110       runs_total      runs_batter  0.973577
86       runs_batter       runs_total  0.973577
46           legbyes      runs_extras  0.635856
94       runs_extras          legbyes  0.635856
20             wides      runs_extras  0.566751
92       runs_extras            wides  0.566751
153  runs_cumulative           inning  0.487396
141           inning  runs_cumulative  0.487396
152  runs_cumulative        over_ball  0.486317
128        over_ball  runs_cumulative  0.486317
59              byes      runs_extras  0.385004
95       runs_extras             byes  0.385004
93       runs_extras          noballs  0.304862
33           noballs      runs_extras  0.304862
106       runs_total          noballs  0.094123

In [13]:
def vis_corr_scatter(x_input, y_input):
    return alt.Chart(corr_data).mark_point().encode(
        x = x_input,
        y = y_input
    ).properties(
        width = 150,
        height = 150
    )
    
batter_total_run_total = vis_corr_scatter("runs_batter:Q", "runs_total:Q")
legbyes_runs_extras = vis_corr_scatter("legbyes:Q", "runs_extras:Q")
wides_runs_extras = vis_corr_scatter("wides:Q", "runs_extras:Q")
inning_runs_cumulative = vis_corr_scatter("inning:Q", "runs_cumulative:Q")	
over_ball_runs_cumulative = vis_corr_scatter("over_ball:Q", "runs_cumulative:Q")
byes_runs_extras = vis_corr_scatter("byes:Q", "runs_extras:Q")
noballs_runs_extras = vis_corr_scatter("noballs:Q", "runs_extras:Q")

v_s_1 = alt.hconcat(batter_total_run_total, legbyes_runs_extras, wides_runs_extras)
v_s_2 = alt.hconcat(inning_runs_cumulative, over_ball_runs_cumulative, byes_runs_extras, noballs_runs_extras)
alt.vconcat(v_s_1, v_s_2)

alt.VConcatChart(...)

Among the quantitative attributes, the top 7 relationships in terms of strengths are(round to 2 decimal places):
1. **batter_total & run_total(r = 0.97)**
2. **legbyes & runs_extras(r = 0.64)**
3. **wides & runs_extra(r = 0.56)**
4. **inning & runs_cumulative(r = 0.49)**
5. **over_ball & runs_cumulative(r = 0.49)**
6. **byes & runs_extra(r = 0.39)**
7. **noballs & run_extra(r = 0.31)**

Note that, using Pearson's r might not be able to capture non-linear relationships that potentially exist among these relationships.